File: backends/ktor/src/main/kotlin/com/hellotimepro/ktor/service/UserService.kt
package com.hellotimepro.ktor.service

import com.hellotimepro.ktor.db.dbQuery
import com.hellotimepro.ktor.domain.User
import com.hellotimepro.ktor.dto.UpdateProfileRequest
import com.hellotimepro.ktor.dto.UserDto
import com.hellotimepro.ktor.repository.UserRepository
import com.hellotimepro.ktor.web.ApiException

class UserService(
    private val users: UserRepository,
    private val avatars: AvatarService,
    private val mapper: MapperService,
) {
    fun toDto(user: User): UserDto = mapper.user(user)

    suspend fun updateProfile(user: User, req: UpdateProfileRequest): UserDto {
        if (req.nickname == null && req.avatarId == null) {
            throw ApiException.validation("至少提供一个字段", "body")
        }
        val newNickname = req.nickname?.let { Validation.nickname(it) }
        val newAvatar = req.avatarId?.let { Validation.avatarFormat(it) }

        return dbQuery {
            var updated = user
            if (newNickname != null && newNickname != user.nickname) {
                if (users.existsByNickname(newNickname)) throw ApiException.conflict("昵称已被使用", "nickname")
                updated = updated.copy(nickname = newNickname)
            }
            if (newAvatar != null) {
                if (!avatars.exists(newAvatar)) throw ApiException.validation("头像 ID 不存在", "avatarId")
                updated = updated.copy(avatarId = newAvatar)
            }
            updated = updated.copy(updatedAt = nowUtc())
            users.update(updated)
            mapper.user(updated)
        }
    }
}

File: backends/aspnet/src/Services/UserService.cs
using HelloTimePro.Aspnet.Domain;
using HelloTimePro.Aspnet.Dto;
using HelloTimePro.Aspnet.Infrastructure;
using HelloTimePro.Aspnet.Repositories;
using HelloTimePro.Aspnet.Web;

namespace HelloTimePro.Aspnet.Services;

public sealed class UserService
{
    private readonly AppDbContext _db;
    private readonly UserRepository _users;
    private readonly AvatarService _avatars;
    private readonly MapperService _mapper;

    public UserService(AppDbContext db, UserRepository users, AvatarService avatars, MapperService mapper)
    {
        _db = db;
        _users = users;
        _avatars = avatars;
        _mapper = mapper;
    }

    public UserDto ToDto(User user) => _mapper.User(user);

    public async Task<UserDto> UpdateProfile(User user, UpdateProfileRequest req)
    {
        if (req.Nickname is null && req.AvatarId is null)
            throw ApiException.Validation("至少提供一个字段", "body");

        var newNickname = req.Nickname is null ? null : Validation.Nickname(req.Nickname);
        var newAvatar = req.AvatarId is null ? null : Validation.AvatarFormat(req.AvatarId);

        if (newNickname is not null && newNickname != user.Nickname)
        {
            if (await _users.ExistsByNickname(newNickname))
                throw ApiException.Conflict("昵称已被使用", "nickname");
            user.Nickname = newNickname;
        }
        if (newAvatar is not null)
        {
            if (!_avatars.Exists(newAvatar))
                throw ApiException.Validation("头像 ID 不存在", "avatarId");
            user.AvatarId = newAvatar;
        }
        user.UpdatedAt = Times.NowUtc();
        await _db.SaveChangesAsync();
        return _mapper.User(user);
    }
}
